### DISCUSSION TITLE
Is it possible to initialise a subclass of a class defined in rust?

### DISCUSSION THREAD

**Discussion by samuelcolvin** [PyO3/pyo3#2729] [Questions]
I assume this is impossible, but since it would be a massive win for pydantic, I thought it worth asking.

Let's say I have:

```rs
#[pyclass(subclass)]
struct MyClass {
    vec: Vec<Whatever>,
}

impl MyClass {
    fn with_capacity() -> Self {
        ....
    }
}

#[pymethods]
impl MyClass {
    #[new]
    fn new() -> Self {
        ...
    }

    pub fn method(&self) -> PyResult<usize> {
        ...
    }
}
```

Because `MyClass` is marked with `#[pyclass(subclass)]` I can subclass it in python.

```py
from my_pyo3_library import MyClass

class MySubclass(MyClass):
    ...
```

Is it possible to pass `MySubclass` back to pyo3/rust code and create instances of `MySubclass` using the `with_capacity` method or similar.

E.g. is it somehow possible to tell rust that I have a variant of `MyClass` with some kind of pointer pointing to it's subclass but that it's a `MyClass` and therefore has a `with_capacity` method?

So something like

```py
from my_pyo3_library import MyClass, do_magic

class MySubclass(MyClass):
    ...

do_magic(MySubclass)
```

```rs
fn do_magic(subclass_of_my_class: &PyAny) {
    let MyClassVariant = magic(subclass_of_my_class)
    MyClassVariant::with_capacity()
```

Or equivilantly, is it possible to create a `MyClass` instance, then cast it to a `MySubclass`?

**Comment by samuelcolvin** [comment]
Ping. Anyone able to give and feedback on this?

Would be great to know even if the answer is "no, you're crazy, that's impossible".

**Reply by birkenfeld** [reply]
I'm not sure what you're asking. `with_capacity` is a Rust-only constructor, so you can't even create proper (Python) instances with it - you'll have to call some method to convert `MyClass` into `Py<MyClass>`.

If the question is more generally how to convert a Rust `MyClass` (or `Py<MyClass>`) into a `PyObject` of type `MySubclass` - IMHO you can't do that just like you can't convert a Python `MyClass` into `MySubclass` (without gross hacks that probably wouldn't work in this case).

So what's left is to create the `MySubclass` from Rust the Python way, by calling the class object, which in turn will call the Rust constructor of `MyClass` at some point.

**Reply by samuelcolvin** [reply]
`MyClass` is accessible in python fine, since it's marked with `#[pyclass(subclass)`, but that's not really the question.

I also don't want to use `with_capacity` in python, but rather in rust, on a subclass of `MyClass`.

**Reply by birkenfeld** [reply]
Both of these points were clear to me when I wrote the response :)

**Reply by samuelcolvin** [reply]
Then how are you not sure what I'm asking? :)

**Reply by birkenfeld** [reply]
Good question. However, a better question, did my reply not help at all?

**Reply by samuelcolvin** [reply]
Well, it made me slightly more confident that the trick I was trying to accomplish is indeed impossible.

But I wanted to be sure you had understood the question before giving up on this, hence trying to clarify what I meant.

**Comment by birkenfeld** [comment]
With the additional info I could glean from your recent comment in #2347, let me have one more go:

It is not possible to create an instance of `MySubclass` in Rust without going through the Python machinery, i.e., calling `MySubclass` using `call()` or similar.

When you have an instance of `MySubclass` in Rust (as a `PyAny` etc) it is possible to get at a `PyRef<MyClass>` using `extract`.

**Reply by samuelcolvin** [reply]
Thanks for this, makes sense.
